arangod/GeneralServer/AsyncJobManager.cpp

```diff
@@ -257,11 +260,36 @@ std::vector<AsyncJobResult::IdType> AsyncJobManager::byStatus(AsyncJobResult::St
   return jobs;
 }
 
+//////////////////////////////////////////////////////////////////////////////
+/// @brief get numbers of pending and done jobs.
+//////////////////////////////////////////////////////////////////////////////
+//
+std::pair<uint64_t, uint64_t> AsyncJobManager::getNrPendingAndDone() {
+  uint64_t pending{0};
+  uint64_t done{0};
+  {
+    READ_LOCKER(readLocker, _lock);
+    for (auto const& j : _jobs) {
+      if (j.second.second._status == AsyncJobResult::JOB_PENDING) {
+        ++pending;
+      } else if (j.second.second._status == AsyncJobResult::JOB_DONE) {
+        ++done;
+      }
+    }
+  }
+  return std::pair(pending, done);
+}
+
 ////////////////////////////////////////////////////////////////////////////////
-/// @brief initializes an async job
+/// @brief initializes an async job, throws if soft shutdown is already
+/// ongoing.
 ////////////////////////////////////////////////////////////////////////////////
 
 void AsyncJobManager::initAsyncJob(std::shared_ptr<RestHandler> handler) {
+  if (_softShutdownOngoing.load(std::memory_order_relaxed)) {
```

Comment:
There is the (rare) possibility for the following sequence of events:

- thread #1 calls `initAsyncJob()` before a soft shutdown is requested, and gets past the initial check for `_softShutdownOngoing` and then goes to sleep.
- thread #2 initiates a soft shutdown, so that `_softShutdownOngoing` now is true.
- thread #2 or any other thread than #1 checks the number of pending operations, which is still zero. this may lead to the conclusion that a soft shutdown is alright, because there is nothing left to execute.
- thread #1 wakes up and inserts the job into the `_jobs` map.

We could prevent this from happening by performing the check on `_softShutdownOngoing` in this method under the write lock.

Reply:
This is a good idea, will do.

Reply:
Done.